zpeclient.go

```diff
@@ -0,0 +1,227 @@
+		if !ok {
+			crMap.roleToPrincipals[string(role.Name)] = []*simplePrincipal{}
+		}
+		if role.RoleMembers == nil {
+			fmt.Println("No role members are found")
+			continue
+		}
+		// Handle trust domains: if string(role.Trust) != ""
+		for _, roleMember := range role.RoleMembers {
+			if roleMember == nil || roleMember.MemberName == "" {
+				continue
+			}
+			memberRegex, err := regexp.Compile("^" + replacer.Replace(strings.ToLower(string(roleMember.MemberName))) + "$")
+			if err != nil {
+				fmt.Printf("Error occurred when converting role memeber name into regex format. Error: %v", err)
+			}
+			principalData := &simplePrincipal{
+				memberRegex: memberRegex,
+			}
+			if roleMember.Expiration == nil {
+				principalData.expiration = time.Time{}
+			} else {
+				principalData.expiration = roleMember.Expiration.Time
+			}
+			crMap.roleToPrincipals[string(role.Name)] = append(crMap.roleToPrincipals[string(role.Name)], principalData)
+		}
+	}
+
+	if item == nil || item.Spec.SignedDomain.Domain == nil || item.Spec.SignedDomain.Domain.Policies == nil || item.Spec.SignedDomain.Domain.Policies.Contents == nil {
+		return errors.New("Some required fields are nil")
+	}
+	for _, policy := range item.Spec.SignedDomain.Domain.Policies.Contents.Policies {
+		if policy == nil {
+			continue
+		}
+		for _, assertion := range policy.Assertions {
+			if assertion == nil {
+				continue
+			}
+			_, ok := crMap.roleToAssertion[assertion.Role]
+			if !ok {
+				crMap.roleToAssertion[assertion.Role] = []*simpleAssertion{}
+			}
+			effect := assertion.Effect.String()
+			resourceRegex, err := regexp.Compile("^" + replacer.Replace(strings.ToLower(assertion.Resource)) + "$")
+			if err != nil {
+				fmt.Printf("Error occurred when converting assertion resource into regex format. Error: %v", err)
+			}
+			actionRegex, err := regexp.Compile("^" + replacer.Replace(strings.ToLower(assertion.Action)) + "$")
+			if err != nil {
+				fmt.Printf("Error occurred when converting assertion action into regex format. Error: %v", err)
+			}
+			simpleAssert := simpleAssertion{
+				resource: resourceRegex,
+				action:   actionRegex,
+				effect:   effect,
+			}
+			crMap.roleToAssertion[assertion.Role] = append(crMap.roleToAssertion[assertion.Role], &simpleAssert)
+		}
+	}
+	return nil
+}
+
+func (c *Cache) addObj(item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := c.DomainMap[domainName]
+	if !ok {
+		roleToPrincipals := make(map[string][]*simplePrincipal)
+		roleToAssertion := make(map[string][]*simpleAssertion)
+		crMap := roleMappings{
+			roleToPrincipals: roleToPrincipals,
+			roleToAssertion:  roleToAssertion,
+		}
+		c.DomainMap[domainName] = crMap
+	}
+	parseData(c.DomainMap, domainName, item)
+	c.lock.Unlock()
+}
+
+func (c *Cache) updateObj(item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := c.DomainMap[domainName]
+	if ok {
+		delete(c.DomainMap, domainName)
```

Comment:
Can we combine add and update together? For example, we can add this if statement to the add function directly and avoid the double lock / unlock. The current approach exposes a section which might be taken by another waiting lock.